File: ofis-gorev-takibi.RETIRED-20260728-145028/src/discord/messages.js
'use strict';

/**
 * Discord mesaj şablonlarının işlenmesi.
 *
 * Şablonlar `scheduled_messages.message_template` sütununda tutulur ve süslü
 * parantezli değişkenler içerir. Böylece yeni mesaj veya görev türü eklerken
 * kod yazmak gerekmez.
 *
 * KULLANILABİLİR DEĞİŞKENLER
 * --------------------------
 *   {name}     -> Çalışanın adı soyadı           (örn. "Beril Kahramanca")
 *   {first}    -> Yalnızca ilk ad                (örn. "Beril")
 *   {mention}  -> Discord etiketi; ID yoksa isim (örn. "<@123...>" veya "Beril Kahramanca")
 *   {emoji}    -> Görev türünün emojisi          (örn. "☕")
 *   {duty}     -> Görev türünün adı              (örn. "Çay")
 *   {company}  -> Ayarlardaki şirket adı
 *   {date}     -> Uzun Türkçe tarih              (örn. "17 Temmuz 2026 Cuma")
 *   {gun}      -> Görev günü, düz hâl            ("Yarın" veya "Pazartesi")
 *   {gunUn}    -> Görev günü, -in hâli           ("Yarının" veya "Pazartesinin")
 *
 * Örnek şablon:
 *   "☀️ Günaydın. Bugünkü çay görevlisi: {emoji} {name}. Herkese iyi çalışmalar."
 */

const { formatLongTR, todayISO, addDays } = require('../utils/date');

/** Hafta günü indeksi (0=Pazar) → düz ve -in hâlli Türkçe ad. */
const WEEKDAY_LABELS = [
  { plain: 'Pazar', genitive: 'Pazarın' },
  { plain: 'Pazartesi', genitive: 'Pazartesinin' },
  { plain: 'Salı', genitive: 'Salının' },
  { plain: 'Çarşamba', genitive: 'Çarşambanın' },
  { plain: 'Perşembe', genitive: 'Perşembenin' },
  { plain: 'Cuma', genitive: 'Cumanın' },
  { plain: 'Cumartesi', genitive: 'Cumartesinin' },
];

/**
 * Bir görev tarihi için gün etiketini üretir.
 *
 * Görev tarihi tam olarak YARIN ise "Yarın" denir (en tanıdık ifade). Değilse —
 * örneğin Cuma 17:00'de sorulan gün Pazartesi'dir, aradaki hafta sonu atlanır —
 * gün adı kullanılır: "Pazartesinin görevlisi belirleniyor".
 *
 * @param {string} targetISO Görev günü (YYYY-MM-DD).
 * @param {string} [todayIso] Karşılaştırma günü; varsayılan bugün.
 * @returns {{ plain: string, genitive: string }}
 */
function dayLabel(targetISO, todayIso = todayISO()) {
  if (targetISO && targetISO === addDays(todayIso, 1)) {
    return { plain: 'Yarın', genitive: 'Yarının' };
  }
  const weekdayIndex = new Date(`${targetISO}T12:00:00Z`).getUTCDay();
  return WEEKDAY_LABELS[weekdayIndex] ?? { plain: 'Yarın', genitive: 'Yarının' };
}

/**
 * Discord kullanıcı etiketi üretir. ID tanımlı değilse düz isme düşer,
 * böylece Discord hesabı olmayan çalışan için de mesaj anlamlı kalır.
 */
function buildMention(employee) {
  return employee.discord_user_id ? `<@${employee.discord_user_id}>` : employee.full_name;
}

/** "Beril Kahramanca" -> "Beril" (samimi hitap için {first} değişkeni). */
function firstName(fullName) {
  return String(fullName ?? '').trim().split(/\s+/)[0] || String(fullName ?? '');
}

/**
 * Şablondaki değişkenleri gerçek değerlerle değiştirir.
 *
 * @param {string} template Ham şablon metni.
 * @param {object} context
 * @param {object} context.employee   Görevli çalışan.
 * @param {object} context.dutyType   Görev türü kaydı.
 * @param {string} [context.company]  Şirket adı.
 * @param {string} [context.date]     ISO tarih; varsayılan bugün.
 * @returns {string} Gönderilmeye hazır mesaj.
 */
function renderTemplate(template, { employee, dutyType, company = '', date = todayISO() }) {
  const day = dayLabel(date);
  const values = {
    name: employee.full_name,
    first: firstName(employee.full_name),
    mention: buildMention(employee),
    emoji: dutyType.emoji || '',
    duty: dutyType.name,
    company,
    date: formatLongTR(date),
    // Gün etiketi: yarınsa "Yarın", değilse gün adı ("Pazartesi"). Cuma 17:00'de
    // sorulan gün Pazartesi olduğu için otomatik "Pazartesi" olur.
    gun: day.plain,
    gunUn: day.genitive, // "Yarının" / "Pazartesinin" (görevlisi belirleniyor)
  };

  // {key} kalıplarını tek geçişte değiştir. Tanınmayan değişken olduğu gibi kalır,
  // böylece şablon hatası sessizce boş metne dönüşmez, gözle fark edilir.
  return template.replace(/\{(\w+)\}/g, (match, key) =>
    Object.prototype.hasOwnProperty.call(values, key) ? values[key] : match
  );
}

module.exports = { renderTemplate, buildMention, firstName, dayLabel };

File: office-portal/apps/office_bot/services/messages.py
"""Discord mesaj sablonlarinin islenmesi — discord/messages.js'in Python portu."""
from __future__ import annotations

from datetime import date

from .date_utils import add_days, format_long_tr, today_iso

WEEKDAY_LABELS = [
    {"plain": "Pazartesi", "genitive": "Pazartesinin"},
    {"plain": "Salı", "genitive": "Salının"},
    {"plain": "Çarşamba", "genitive": "Çarşambanın"},
    {"plain": "Perşembe", "genitive": "Perşembenin"},
    {"plain": "Cuma", "genitive": "Cumanın"},
    {"plain": "Cumartesi", "genitive": "Cumartesinin"},
    {"plain": "Pazar", "genitive": "Pazarın"},
]


def day_label(target_iso: str, today: str | None = None):
    today = today or today_iso()
    if target_iso and target_iso == add_days(today, 1):
        return {"plain": "Yarın", "genitive": "Yarının"}
    weekday = date.fromisoformat(target_iso).weekday()  # Pazartesi=0
    return WEEKDAY_LABELS[weekday]


def build_mention(employee) -> str:
    discord_id = getattr(employee, "discord_user_id", None) or (employee.get("discord_user_id") if isinstance(employee, dict) else None)
    full_name = getattr(employee, "full_name", None) or (employee.get("full_name") if isinstance(employee, dict) else "")
    return f"<@{discord_id}>" if discord_id else full_name


def first_name(full_name: str) -> str:
    parts = str(full_name or "").strip().split()
    return parts[0] if parts else str(full_name or "")


def render_template(template: str, *, employee, duty_type, company: str = "", target_date: str | None = None) -> str:
    target_date = target_date or today_iso()
    day = day_label(target_date)
    full_name = getattr(employee, "full_name", None) or (employee.get("full_name") if isinstance(employee, dict) else "")
    emoji = getattr(duty_type, "emoji", None) or (duty_type.get("emoji") if isinstance(duty_type, dict) else "")
    duty_name = getattr(duty_type, "name", None) or (duty_type.get("name") if isinstance(duty_type, dict) else "")

    values = {
        "name": full_name,
        "first": first_name(full_name),
        "mention": build_mention(employee),
        "emoji": emoji or "",
        "duty": duty_name,
        "company": company,
        "date": format_long_tr(target_date),
        "gun": day["plain"],
        "gunUn": day["genitive"],
    }

    import re

    return re.sub(r"\{(\w+)\}", lambda m: str(values.get(m.group(1), m.group(0))), template)
